iree/hal/allocator_caching.c

```diff
@@ -0,0 +1,220 @@
+  cache_buffer_node->buffer = buffer;
+  cache_buffer_node->next = NULL;
+
+  if(iree_hal_caching_allocator_cache_list == NULL) {
+    iree_hal_caching_allocator_cache_list = cache_buffer_node;
+  } else {
+    iree_hal_caching_allocator_cache
+      *iree_hal_caching_allocator_cache_ptr = iree_hal_caching_allocator_cache_list;
+    // Add buffer to the end of list
+    while(iree_hal_caching_allocator_cache_ptr != NULL) {
+      if(iree_hal_caching_allocator_cache_ptr->next == NULL)
+        break;
+      else
+        iree_hal_caching_allocator_cache_ptr = iree_hal_caching_allocator_cache_ptr->next;
+    }
+    iree_hal_caching_allocator_cache_ptr->next = cache_buffer_node;
+  }
+  return iree_ok_status();
+}
+
+static void iree_hal_caching_allocator_destroy(
+    iree_hal_allocator_t* base_allocator) {
+  iree_hal_allocator_cache_buffer = false;
+  iree_hal_caching_allocator_t* allocator = iree_hal_caching_allocator_cast(base_allocator);
+  iree_hal_allocator_destroy(allocator->delegate_allocator);
+}
+
+static iree_allocator_t iree_hal_caching_allocator_host_allocator(
+    const iree_hal_allocator_t* base_allocator) {
+  
+  return iree_hal_allocator_host_allocator(base_allocator);
+}
+
+static void iree_hal_caching_allocator_query_statistics(
+    iree_hal_allocator_t* base_allocator,
+    iree_hal_allocator_statistics_t* out_statistics) {
+  iree_hal_caching_allocator_t* allocator = iree_hal_caching_allocator_cast(base_allocator);
+  iree_hal_allocator_query_statistics(allocator->delegate_allocator, out_statistics);
+}
+
+static iree_hal_buffer_compatibility_t
+iree_hal_caching_allocator_query_buffer_compatibility(
+    iree_hal_allocator_t* base_allocator, iree_hal_memory_type_t memory_type,
+    iree_hal_buffer_usage_t allowed_usage,
+    iree_hal_buffer_usage_t intended_usage,
+    iree_device_size_t allocation_size) {
+
+  iree_hal_caching_allocator_t* allocator = iree_hal_caching_allocator_cast(base_allocator);
+  return iree_hal_allocator_query_buffer_compatibility(allocator->delegate_allocator,
+    memory_type, allowed_usage, intended_usage, allocation_size);
+}
+
+static iree_status_t iree_hal_caching_allocator_allocate_buffer(
+    iree_hal_allocator_t* base_allocator, iree_hal_memory_type_t memory_type,
+    iree_hal_buffer_usage_t allowed_usage, iree_host_size_t allocation_size,
+    iree_hal_buffer_t** out_buffer) {
+
+  iree_hal_caching_allocator_t* allocator = iree_hal_caching_allocator_cast(base_allocator);
+
+  if (iree_all_bits_set(memory_type, IREE_HAL_MEMORY_TYPE_DEVICE_LOCAL)) {
+    if (iree_all_bits_set(memory_type, IREE_HAL_MEMORY_TYPE_HOST_VISIBLE)) {
+      if(iree_hal_caching_allocator_buffer_available(allocation_size)) {
+          return iree_hal_caching_allocator_allocate_from_cache(allocation_size, out_buffer);
+      }
+    }
+  }
+
+  return iree_hal_allocator_allocate_buffer(allocator->delegate_allocator, memory_type,
```

Comment:
I think you are also going to want to create a custom buffer wrapper instead of just returning the delegate buffer directly. The reason is that then you can provide a vtable with a custom destroy method that returns it to this cache (this would remove the special case you currently have in cuda_buffer.c.

See cuda_buffer.c:46 for how to allocate and return a custom/wrapping buffer resource. In its destroy hook, add it back to the cache.

Reply:
I created a custom buffer (intrusive list) that stores a pointer to the underlying buffer as well as a resource. Like you mentioned the idea is to have the destroy hook add buffer back into the cache. However, I run into buffer incompatibility issue when running an example.  
```
PERMISSION_DENIED; buffer memory type is not compatible with the requested operation; buffer has , operation requires HOST_VISIBLE; parsing value '1x5xf32=1,-2,-3,4,-5'
```

As for methods other than destroy in the vtable, I try to invoke the underlying vtable of delegate buffer but I find that the method signature in vtable (https://github.com/google/iree/blob/main/iree/hal/buffer.h#L480) is different from the top level call signature (https://github.com/google/iree/blob/main/iree/hal/buffer.h#L411). I'm trying to understand this better but any pointer from you would be really helpful.

Sorry to come to you with an error, I've been trying to hack around this but haven't been successful.

Reply:
You are going to need to wrap any such buffers and preserve the mapping_mode. It is hard to spot the exact problem without new code pushed. Can you update the branch with your pending changes?